draft-normington-jsonpath-latest.xml

```diff
@@ -384,50 +384,115 @@ array-slice = [ integer ] ws %x3A ws [ integer ]
+
+                     <t>A slice expression selects a subset of the elements of the input array, in the same order
+                        as the array or the reverse order, depending on the sign of the <tt>step</tt> parameter.</t>
+
+                     <t>A slice is defined by the two slice parameters, <tt>start</tt> and <tt>end</tt>,
+                        and an iteration delta, <tt>step</tt>. Each of these parameters is optional.
+                        <tt>len</tt> is the length of the input array.</t>
+
+                     <t>The default value for <tt>step</tt> is <tt>1</tt>.
+                        The default values for <tt>start</tt> and <tt>end</tt> depend on the sign of <tt>step</tt>,
+                        as defined in the following pseudocode:</t>
+
+                     <sourcecode type="pseudocode">
+IF step &gt;= 0 THEN
+
+  IF start is omitted THEN
+    start = 0
+  END IF
+  IF end is omitted THEN
+    end = len
+  END IF
+
+ELSE
+
+  IF start is omitted THEN
+    start = len - 1
+  END IF
+  IF end is omitted THEN
+    end = -len - 1
+  END IF
+
+END IF
                      </sourcecode>
-                     <t>When <tt>step</tt> is positive, the default value of <tt>start</tt> is <tt>0</tt> and the default value of <tt>end</tt> is the length of the array.
-                     </t>
-                     <t>For negative integer <tt>step</tt> and integers <tt>start</tt> and <tt>end</tt>, the array slice
-                        <tt>start:end:step</tt> consists of the indices starting at <tt>start</tt> and decreasing repeatedly
-                        by <tt>step</tt> while the resultant index is greater than <tt>end</tt>.
-                        The indices of the slice are the same as the values of i iterated over by this <tt>for</tt> loop in C:
-                     </t>
-                     <sourcecode type="c">
-for (i = start; i &gt; end; i = i + step) {
-  ...
-}
+
+                     <t>Slice expression parameters <tt>start</tt> and <tt>end</tt> are not directly usable
+                        as slice bounds and must be first normalized. Normalization is defined as:</t>
+
+                     <sourcecode type="pseudocode">
+FUNCTION Normalize(i):
+  IF i &gt;= 0 THEN
+    RETURN i
+  ELSE
+    RETURN len + i
+  END IF
                      </sourcecode>
-                     <t>When <tt>step</tt> is negative, the default value of <tt>start</tt> is <tt>len - 1</tt> (where <tt>len</tt> is the length of the array)
-                        and the default value of <tt>end</tt> is <tt>-1</tt> (literally, not in the earlier sense of a shorthand for <tt>len - 1</tt>).
-                     </t>
-                     <t>For zero <tt>step</tt>, the array slice is empty and no elements are selected.
-                     </t>
+
+                     <t>The semantics of the array indexing expression <tt>[i]</tt> is defined as syntactic sugar for
+                        <tt>a[i:Normalize(i)+1:1]</tt>.</t>
```

Comment:
It seems strange to talk about _syntactic_ sugar in the semantics section. Perhaps like the following would be clearer:
```suggestion
                     <t>The result of the array indexing expression <tt>[i]</tt> is defined to be the result of the array slicing expression
                        <tt>a[i:Normalize(i)+1:1]</tt>.</t>
```